libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs

```diff
@@ -0,0 +1,217 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Cognitive.LUIS;
+using Microsoft.Cognitive.LUIS.Models;
+using Newtonsoft.Json.Linq;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Bot.Builder.LUIS
+{
+    /// <inheritdoc />
+    /// <summary>
+    /// A Luis based implementation of IRecognizer
+    /// </summary>
+    public class LuisRecognizer : IRecognizer
+    {
+        private readonly LuisService _luisService;
+        private readonly ILuisOptions _luisOptions;
+        private readonly ILuisRecognizerOptions _luisRecognizerOptions;
+
+        public LuisRecognizer(ILuisModel luisModel, ILuisRecognizerOptions luisRecognizerOptions = null, ILuisOptions options = null)
+        {
+            _luisService = new LuisService(luisModel);
+            _luisOptions = options;
+            _luisRecognizerOptions = luisRecognizerOptions;
+        }
+
+        /// <inheritdoc />
+        public Task<RecognizerResult> Recognize(string utterance, CancellationToken ct)
+        {
+            var luisRequest = new LuisRequest(utterance);
+            _luisOptions?.Apply(luisRequest);
+            return Recognize(luisRequest, ct, _luisRecognizerOptions?.Verbose ?? true);
+        }
+
+        private async Task<RecognizerResult> Recognize(LuisRequest request, CancellationToken ct, bool verbose)
+        {
+            var luisResult = await _luisService.QueryAsync(request, ct);
+
+            var recognizerResult = new RecognizerResult
+            {
+                Text = request.Query,
+                Intents = GetIntents(luisResult),
+                Entities = GetEntitiesAndMetadata(luisResult.Entities, luisResult.CompositeEntities, verbose)
+            };
+
+            return recognizerResult;
+        }
+
+        private static JObject GetIntents(LuisResult luisResult)
+        {
+            return luisResult.Intents != null ?
+                JObject.FromObject(luisResult.Intents.ToDictionary(i => i.Intent, i => i.Score ?? 0)) :
+                new JObject { [luisResult.TopScoringIntent.Intent] = luisResult.TopScoringIntent.Score ?? 0 };
+        }
+
+        private static JObject GetEntitiesAndMetadata(IList<EntityRecommendation> entities, IList<CompositeEntity> compositeEntities, bool verbose)
+        {
+            var entitiesAndMetadata = new JObject();
+            if (verbose)
+            {
+                entitiesAndMetadata["$instance"] = new JObject();
+            }
+            var compositeEntityTypes = new HashSet<string>();
```

Comment:
Why a HashSet? List<string> seems like it would do the trick, and be a little more clear. 

(Semantics are the same, but anytime I see HashSet, I wonder "Why did they pick this?". In this case, it doesn't seem specific to the problem domain, so just List<T> would be more clear...)

Reply:
HashSet to make look ups fast (O(1)), some LUIS models have hundreds of entities.